ic-utils/src/interfaces/wallet.rs

```diff
@@ -0,0 +1,184 @@
+            .build())
+    }
+
+    pub async fn call(self) -> Result<RequestId, AgentError> {
+        self.build()?.call().await
+    }
+
+    pub async fn call_and_wait<W>(self, waiter: W) -> Result<Out, AgentError>
+    where
+        Out: 'agent + for<'de> ArgumentDecoder<'de> + Send + Sync,
+        W: Waiter,
+    {
+        self.build()?.call_and_wait(waiter).await
+    }
+}
+
+#[async_trait]
+impl<'agent, 'canister: 'agent, Out> AsyncCall<Out> for CallForward<'agent, 'canister, Out>
+where
+    Out: for<'de> ArgumentDecoder<'de> + Send + Sync,
+{
+    async fn call(self) -> Result<RequestId, AgentError> {
+        self.call().await
+    }
+
+    async fn call_and_wait<W>(self, waiter: W) -> Result<Out, AgentError>
+    where
+        W: Waiter,
+    {
+        self.call_and_wait(waiter).await
+    }
+}
+
+/// A wallet canister interface, for the standard wallet provided by DFINITY.
+/// This interface implement most methods conveniently for the user.
+pub struct Wallet;
+
+impl Wallet {
+    /// Create an instance of a [Canister] implementing the Wallet interface
+    /// and pointing to the right Canister ID.
+    pub fn create(agent: &Agent, canister_id: Principal) -> Canister<Wallet> {
+        Canister::builder()
+            .with_agent(agent)
+            .with_canister_id(canister_id)
+            .with_interface(Wallet)
+            .build()
+            .unwrap()
+    }
+
+    /// Creating a CanisterBuilder with the right interface and Canister Id. This can
+    /// be useful, for example, for providing additional Builder information.
+    pub fn with_agent(agent: &Agent) -> CanisterBuilder<Wallet> {
+        Canister::builder().with_agent(agent).with_interface(Wallet)
+    }
+}
+
+impl<'agent> Canister<'agent, Wallet> {
+    /// Get the current controller's principal ID.
+    pub fn get_controller<'canister: 'agent>(
+        &'canister self,
+    ) -> impl 'agent + SyncCall<(Principal,)> {
+        self.query_("get_controller").build()
+    }
+
+    /// Transfer controller to another principal ID.
+    pub fn set_controller<'canister: 'agent>(
+        &'canister self,
+        principal: Principal,
+    ) -> impl 'agent + AsyncCall<()> {
+        self.update_("set_controller").with_arg(principal).build()
+    }
+
+    /// Get the list of custodians.
+    pub fn get_custodians<'canister: 'agent>(
+        &'canister self,
+    ) -> impl 'agent + SyncCall<(Vec<Principal>,)> {
+        self.query_("get_custodians").build()
+    }
+
+    /// Authorize a new custodian.
+    pub fn authorize<'canister: 'agent>(
+        &'canister self,
+        custodian: Principal,
+    ) -> impl 'agent + AsyncCall<()> {
+        self.update_("authorize").with_arg(custodian).build()
+    }
+
+    /// Deauthorize a custodian.
+    pub fn deauthorize<'canister: 'agent>(
+        &'canister self,
+        custodian: Principal,
+    ) -> impl 'agent + AsyncCall<()> {
+        self.update_("authorize").with_arg(custodian).build()
```

Comment:
should this be the following?
```suggestion
        self.update_("deauthorize").with_arg(custodian).build()
```